### ISSUE TITLE
Unexpected OBJECT_PATTERN parent

### ISSUE THREAD

**Issue by jplaisted** [google/closure-compiler#2061]
Messing around with object destructing and I got a compilation error + stack trace! :(

Note that this is valid JS - you can run it in Chrome.

``` js
function readAll(reader) { 
    let values = [];
    let value;
    let done;
    while (({value, done} = reader.read()) && !done) {
      values.push(value);
    }
    return values;
}

class Reader { 
  constructor(arr) { 
    this.arr = arr; 
  } 

  read() { 
    return this.arr.length ? 
      {value: this.arr.shift(), done: false} : 
      {done: true}; 
  } 
}

readAll(new Reader([1,2,4,5]));
```

```
23: java.lang.IllegalStateException: Unexpected OBJECT_PATTERN parent: ASSIGN 5 [length: 29] [source_file: Input_0]
    at com.google.javascript.jscomp.Es6RewriteDestructuring.visitObjectPattern(Es6RewriteDestructuring.java:224)
    at com.google.javascript.jscomp.Es6RewriteDestructuring.visit(Es6RewriteDestructuring.java:78)
    at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:650)
    at com.google.javascript.jscomp.NodeTraversal.traverseChildren(NodeTraversal.java:722)
    ...
```

**Comment by MatrixFrog** [comment]
https://closure-compiler-debugger.appspot.com/#input0%3Dfunction%2520readAll(reader)%2520%257B%2520%250A%2520%2520%2520%2520let%2520value%253B%250A%2520%2520%2520%2520let%2520done%253B%250A%2520%2520%2520%2520while%2520(%257Bvalue%252C%2520done%257D%2520%253D%2520reader.read())%2520%257B%250A%2520%2520%2520%2520%257D%250A%257D%250A%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_SYMBOLS%3D1%26CHECK_TYPES%3D1%26CLOSURE_PASS%3D1%26LANG_IN_IS_ES6%3D1%26MISSING_PROPERTIES%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26PRETTY_PRINT%3D1%26TRANSPILE%3D1

**Comment by MatrixFrog** [comment]
It looks like we handle standalone ASSIGN statements, but not ASSIGN expressions nested inside other expressions (like a while or if). Good catch.

**Comment by jaubourg** [comment]
Any update on this one ? Just bumped into it today ;)

**Comment by MatrixFrog** [comment]
No update, sorry. It probably shouldn't take too long, we just haven't prioritized it yet.

**Comment by lauraharker** [comment]
This seems to be fixed in 96ff594.

https://closure-compiler-debugger.appspot.com/#input0%3Dfunction%2520readAll(reader)%2520%257B%2520%250A%2520%2520%2520%2520let%2520value%253B%250A%2520%2520%2520%2520let%2520done%253B%250A%2520%2520%2520%2520while%2520(%257Bvalue%252C%2520done%257D%2520%253D%2520reader.read())%2520%257B%250A%2520%2520%2520%2520%257D%250A%257D%250A%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_SYMBOLS%3D1%26CHECK_TYPES%3D1%26CLOSURE_PASS%3D1%26LANG_IN_IS_ES6%3D1%26MISSING_PROPERTIES%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26PRETTY_PRINT%3D1%26TRANSPILE%3D1
